File: android/app/src/main/kotlin/com/menumaker/data/repository/IntegrationRepository.kt
package com.menumaker.data.repository

import com.menumaker.data.common.Resource
import com.menumaker.data.remote.api.ApiService
import com.menumaker.data.remote.models.IntegrationDto
import com.menumaker.data.remote.models.PaymentProcessorData
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import javax.inject.Inject

interface IntegrationRepository {
    fun getIntegrations(businessId: String): Flow<Resource<List<IntegrationDto>>>
    fun connectPOS(provider: String): Flow<Resource<PaymentProcessorData>>
    fun connectDelivery(provider: String): Flow<Resource<PaymentProcessorData>>
    fun disconnectIntegration(id: String): Flow<Resource<Unit>>
}

class IntegrationRepositoryImpl @Inject constructor(
    private val apiService: ApiService
) : IntegrationRepository {

    private val launchGatedIntegrationTypes = setOf("pos", "delivery")
    private val unsafeTextControls =
        Regex("[\\u0000-\\u0008\\u000B\\u000C\\u000E-\\u001F\\u007F-\\u009F\\u200B-\\u200F\\u202A-\\u202E\\u2060-\\u206F\\uFEFF]")

    private fun launchGatedMessage(capability: String, provider: String): String =
        "$capability provider '$provider' is launch-gated in this build and remains disabled until backend capability evidence is recorded."

    private fun normalizeBoundaryText(label: String, value: String, maxLength: Int = 255): String {
        val normalized = value.trim()
        require(normalized.isNotEmpty()) { "$label is required" }
        require(normalized.length <= maxLength) { "$label must be $maxLength characters or fewer" }
        require(!unsafeTextControls.containsMatchIn(normalized)) {
            "$label contains unsafe control characters"
        }
        return normalized
    }

    private fun isVisibleIntegrationBoundaryText(label: String, value: String, maxLength: Int): Boolean =
        try {
            normalizeBoundaryText(label, value, maxLength)
            true
        } catch (_: IllegalArgumentException) {
            false
        }

    private fun visibleIntegrationsWhileLaunchGated(integrations: List<IntegrationDto>): List<IntegrationDto> =
        integrations.filterNot { integration ->
            !isVisibleIntegrationBoundaryText("Integration ID", integration.id, 255) ||
                !isVisibleIntegrationBoundaryText("Integration provider", integration.provider, 64) ||
                !isVisibleIntegrationBoundaryText("Integration type", integration.type, 64) ||
                integration.type.trim().lowercase() in launchGatedIntegrationTypes
        }

    override fun getIntegrations(businessId: String): Flow<Resource<List<IntegrationDto>>> = flow {
        emit(Resource.Loading)
        try {
            val normalizedBusinessId = normalizeBoundaryText("Business ID", businessId)
            val response = apiService.getIntegrations(normalizedBusinessId)
            if (response.isSuccessful && response.body() != null) {
                val integrations = visibleIntegrationsWhileLaunchGated(response.body()!!.data.integrations)
                emit(Resource.Success(integrations))
            } else {
                emit(Resource.Error(response.message() ?: "Failed to load integrations"))
            }
        } catch (e: Exception) {
            emit(Resource.Error(e.message ?: "An error occurred", e))
        }
    }

    override fun connectPOS(provider: String): Flow<Resource<PaymentProcessorData>> = flow {
        emit(Resource.Loading)
        try {
            emit(Resource.Error(launchGatedMessage("POS", normalizeBoundaryText("POS provider", provider))))
        } catch (e: IllegalArgumentException) {
            emit(Resource.Error(e.message ?: "Invalid POS provider", e))
        }
    }

    override fun connectDelivery(provider: String): Flow<Resource<PaymentProcessorData>> = flow {
        emit(Resource.Loading)
        try {
            emit(Resource.Error(launchGatedMessage("Delivery", normalizeBoundaryText("Delivery provider", provider))))
        } catch (e: IllegalArgumentException) {
            emit(Resource.Error(e.message ?: "Invalid delivery provider", e))
        }
    }

    override fun disconnectIntegration(id: String): Flow<Resource<Unit>> = flow {
        emit(Resource.Loading)
        try {
            val normalizedId = normalizeBoundaryText("Integration ID", id)
            emit(Resource.Error("Integration disconnect for '$normalizedId' is launch-gated in this build because POS and delivery-provider integrations remain disabled until backend capability evidence is recorded."))
        } catch (e: IllegalArgumentException) {
            emit(Resource.Error(e.message ?: "Invalid integration ID", e))
        }
    }
}

File: ios/MenuMaker/Data/Repositories/IntegrationRepository.swift
import Foundation
import Combine

private enum IntegrationCapabilityError: LocalizedError {
    case launchGated(String)

    var errorDescription: String? {
        switch self {
        case .launchGated(let capability):
            return "\(capability) integrations are launch-gated in this build and remain disabled until backend capability evidence is recorded."
        }
    }
}

/// Integration repository
@MainActor
class IntegrationRepository: ObservableObject {
    static let shared = IntegrationRepository()

    private let apiClient = APIClient.shared
    private static let launchGatedIntegrationTypes: Set<String> = ["pos", "delivery"]
    private static let unsafeTextScalarValues: Set<UInt32> = {
        Set((Array(0x0000...0x0008)
            + [0x000B, 0x000C]
            + Array(0x000E...0x001F)
            + Array(0x007F...0x009F)
            + Array(0x200B...0x200F)
            + Array(0x202A...0x202E)
            + Array(0x2060...0x206F)
            + [0xFEFF]).map(UInt32.init))
    }()
    private static let queryValueAllowed = CharacterSet(charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")

    @Published var integrations: [Integration] = []

    private init() {}

    enum BoundaryError: LocalizedError, Equatable {
        case required(String)
        case tooLong(String, Int)
        case unsafeControlCharacters(String)

        var errorDescription: String? {
            switch self {
            case .required(let label):
                return "\(label) is required"
            case .tooLong(let label, let maxLength):
                return "\(label) must be \(maxLength) characters or fewer"
            case .unsafeControlCharacters(let label):
                return "\(label) contains unsafe control characters"
            }
        }
    }

    static func normalizeIntegrationText(_ label: String, _ value: String, maxLength: Int) throws -> String {
        guard !containsUnsafeTextScalar(value) else {
            throw BoundaryError.unsafeControlCharacters(label)
        }
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty else {
            throw BoundaryError.required(label)
        }
        guard normalized.count <= maxLength else {
            throw BoundaryError.tooLong(label, maxLength)
        }
        return normalized
    }

    private static func containsUnsafeTextScalar(_ value: String) -> Bool {
        value.unicodeScalars.contains { unsafeTextScalarValues.contains($0.value) }
    }

    static func integrationListEndpoint(for businessId: String) throws -> String {
        let normalizedBusinessId = try normalizeIntegrationText("Business ID", businessId, maxLength: 255)
        let encodedBusinessId = normalizedBusinessId.addingPercentEncoding(withAllowedCharacters: queryValueAllowed) ?? normalizedBusinessId
        return AppConstants.API.Endpoints.integrations + "?business_id=\(encodedBusinessId)"
    }

    // MARK: - Fetch Operations

    func getIntegrations(_ businessId: String) async throws -> [Integration] {
        let response: IntegrationListResponse = try await apiClient.request(
            endpoint: try Self.integrationListEndpoint(for: businessId),
            method: .get
        )

        let visibleIntegrations = Self.visibleIntegrationsWhileLaunchGated(response.data.integrations)
        integrations = visibleIntegrations
        return visibleIntegrations
    }

    static func visibleIntegrationsWhileLaunchGated(_ integrations: [Integration]) -> [Integration] {
        integrations.filter { integration in
            guard (try? normalizeIntegrationText("Integration ID", integration.id, maxLength: 255)) != nil else {
                return false
            }
            guard let normalizedProvider = try? normalizeIntegrationText("Integration provider", integration.provider, maxLength: 64),
                  normalizedProvider.unicodeScalars.allSatisfy({ $0.value >= 0x20 && $0.value <= 0x7E }) else {
                return false
            }
            guard let normalizedType = try? normalizeIntegrationText("Integration type", integration.type, maxLength: 64).lowercased() else {
                return false
            }
            return !launchGatedIntegrationTypes.contains(normalizedType)
        }
    }

    // MARK: - Connect Operations

    func connectPOS(provider: IntegrationProvider, credentials: [String: String]) async throws -> Integration {
        throw IntegrationCapabilityError.launchGated("POS")
    }

    func connectDelivery(provider: IntegrationProvider, credentials: [String: String]) async throws -> Integration {
        throw IntegrationCapabilityError.launchGated("Delivery-provider")
    }

    // MARK: - Disconnect Operations

    func disconnectIntegration(_ integrationId: String) async throws {
        throw IntegrationCapabilityError.launchGated("POS and delivery-provider")
    }

    // MARK: - Helpers

    func getIntegrationsByType(_ type: IntegrationType) -> [Integration] {
        integrations.filter {
            (try? Self.normalizeIntegrationText("Integration type", $0.type, maxLength: 64).lowercased()) == type.rawValue
        }
    }

    func getPOSIntegrations() -> [Integration] {
        getIntegrationsByType(.pos)
    }

    func getDeliveryIntegrations() -> [Integration] {
        getIntegrationsByType(.delivery)
    }

    func getActiveIntegrations() -> [Integration] {
        integrations.filter { $0.isActive }
    }

    func hasIntegration(provider: IntegrationProvider, type: IntegrationType) -> Bool {
        integrations.contains {
            (try? Self.normalizeIntegrationText("Integration provider", $0.provider, maxLength: 64).lowercased()) == provider.rawValue &&
                (try? Self.normalizeIntegrationText("Integration type", $0.type, maxLength: 64).lowercased()) == type.rawValue
        }
    }

    func getLastSyncTime(for integrationId: String) -> String? {
        guard let normalizedIntegrationId = try? Self.normalizeIntegrationText("Integration ID", integrationId, maxLength: 255) else {
            return nil
        }
        return integrations.first {
            (try? Self.normalizeIntegrationText("Integration ID", $0.id, maxLength: 255)) == normalizedIntegrationId
        }?.formattedLastSync
    }
}
